File: csharp/pascals-triangle/PascalsTriangle.cs
using System;
using System.Collections.Generic;

public static class PascalsTriangle
{
    public static IEnumerable<IEnumerable<int>> Calculate(int rows)
    {
        List<List<int>> tri = new List<List<int>>();
        for (int i = 0; i < rows; i++)
        {
            tri.Add(new List<int>());
            for (int j = 0; j < i + 1; j++)
            {
                tri[i].Add(CalcValue(i, j, tri));
            }
        }
        return tri;
    }

    private static int CalcValue(int row, int el, List<List<int>> tri)
    {
        if (row == 0 && el == 0)
        {
            return 1;
        }

        if (row - 1 == 0)
        {
            return 1;
        }

        if (el == 0 || el == row)
        {
            return 1;
        }
        
        return tri[row - 1][el - 1] + tri[row - 1][el];
    }
}

File: cpp/pascals-triangle/pascals_triangle.cpp
#include "pascals_triangle.h"

namespace pascals_triangle
{

    static int value(int i, int j, std::vector<std::vector<int>> &tri)
    {
        if (i == 0 && j == 0)
            return 1;

        if (i - 1 == 0)
            return 1;

        if (j == 0 || i == j)
            return 1;

        return tri[i - 1][j - 1] + tri[i - 1][j];
    }

    std::vector<std::vector<int>> generate_rows(int rows)
    {

        std::vector<std::vector<int>> result{};

        for (int i = 0; i < rows; i++)
        {

            std::vector<int> row{};
            for (int j = 0; j < i + 1; j++)
            {
                row.push_back(value(i, j, result));
            }
            result.push_back(row);
        }

        return result;
    }

} // namespace pascals_triangle
